api/v1/client.go

```diff
@@ -7,83 +7,96 @@ import (
+	storage "github.com/oasisprotocol/oasis-indexer/storage/client"
 )
 
 const (
-	tpsWindowSizeMinutes = 5
+	blockCost = 1
+	txCost    = 1
 )
 
 // storageClient is a wrapper around a storage.TargetStorage
 // with knowledge of network semantics.
 type storageClient struct {
-	db     storage.TargetStorage
+	storage *storage.StorageClient
+
+	blockCache *ristretto.Cache
+	txCache    *ristretto.Cache
+
 	logger *log.Logger
 }
 
 // newStorageClient creates a new storage client.
-func newStorageClient(db storage.TargetStorage, l *log.Logger) *storageClient {
-	return &storageClient{db, l}
+func newStorageClient(s *storage.StorageClient, l *log.Logger) (*storageClient, error) {
+	blockCache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters:        1024 * 10,
+		MaxCost:            1024,
+		BufferItems:        64,
+		IgnoreInternalCost: true,
+	})
+	if err != nil {
+		l.Error("api client: failed to create block cache: %w", err)
+		return nil, err
+	}
+	txCache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters:        1024 * 10,
+		MaxCost:            1024,
+		BufferItems:        64,
+		IgnoreInternalCost: true,
+	})
+	if err != nil {
+		l.Error("api client: failed to create tx cache: %w", err)
+		return nil, err
+	}
+	return &storageClient{s, blockCache, txCache, l}, nil
 }
 
 // Status returns status information for the Oasis Indexer.
-func (c *storageClient) Status(ctx context.Context) (*Status, error) {
-	qf := NewQueryFactory(strcase.ToSnake(LatestChainID))
-
-	s := Status{
-		LatestChainID: LatestChainID,
-	}
-	if err := c.db.QueryRow(
-		ctx,
-		qf.StatusQuery(),
-	).Scan(&s.LatestBlock, &s.LatestUpdate); err != nil {
-		c.logger.Info("row scan failed",
-			"request_id", ctx.Value(RequestIDContextKey),
-			"err", err.Error(),
-		)
-		return nil, common.ErrStorageError
-	}
-	return &s, nil
+func (c *storageClient) Status(ctx context.Context) (*storage.Status, error) {
+	return c.storage.Status(ctx)
 }
 
 // Blocks returns a list of consensus blocks.
-func (c *storageClient) Blocks(ctx context.Context, r *http.Request) (*BlockList, error) {
-	cid, ok := ctx.Value(ChainIDContextKey).(string)
-	if !ok {
-		return nil, common.ErrBadChainID
-	}
-	qf := NewQueryFactory(cid)
-
+func (c *storageClient) Blocks(ctx context.Context, r *http.Request) (*storage.BlockList, error) {
+	var q storage.BlocksRequest
 	params := r.URL.Query()
-
-	var from *string
 	if v := params.Get("from"); v != "" {
-		from = &v
+		from, err := strconv.ParseInt(v, 10, 64)
+		if err != nil {
+			return nil, common.ErrBadRequest
+		}
+		q.From = &from
 	}
-	var to *string
 	if v := params.Get("to"); v != "" {
-		to = &v
+		to, err := strconv.ParseInt(v, 10, 64)
+		if err != nil {
+			return nil, common.ErrBadRequest
+		}
+		q.To = &to
 	}
-	var after *string
 	if v := params.Get("after"); v != "" {
-		after = &v
+		after, err := time.Parse("2006-01-02T15:04:05Z-0700", v)
+		if err != nil {
+			return nil, common.ErrBadRequest
+		}
```

Comment:
since the layout is repeated, maybe factor this (and the `ParseInt`s above) into a private `validateDatetime` and `validateInt` func respectively. doesn't really reduce LOC but will make any future changes much easier
```suggestion
		after, err := time.Parse("2006-01-02T15:04:05Z-0700", v)
		if err != nil {
			return nil, common.ErrBadRequest
		}
```

Reply:
Updated